### ISSUE TITLE
Calling a method with a dynamic value

### ISSUE THREAD

**Issue by Stamo-Gochev** [microsoft/dotnet#1044]
> This repo is not an official .NET Framework support location, however, we will respond to issues filed here as best we can. 

Note: feel free to move this to another place according to the above, but I hope I can get some details regarding the issue.

I am encountering a strange problem using the following configuration:
```
    public class Program
    {
        public static void Main(string[] args)
        {
            dynamic processed = true;
            var dataProcessor = new TemplateDataProcessor<string>();
            Console.WriteLine(dataProcessor.Equals(processed));
        }
    }

    public class TemplateData<T> { }
    public abstract class DataProcessorBase<TData, TDataBuilder> { }
    public abstract class DataProcessor<TData, TDataBuilder> : DataProcessorBase<TData, TDataBuilder> { }
    public class TemplateDataProcessor<T> : DataProcessor<TemplateData<T>, TemplateDataProcessor<T>> { }
```
This fails with a stackoverflow exception when tested in [.NET 4.7.2](https://dotnetfiddle.net/TAnJge) but works in .NET Core 3.0 (also with 2.2).

If the inheritance is changed to using the `DataProcessorBase` abstract class though:
```
 public class TemplateDataProcessor<T> : DataProcessorBase<TemplateData<T>, TemplateDataProcessor<T>> { }
```
this works in .NET 4.7.2. An

**Comment by sujayvsarma** [comment]
From what I can see it has something to do with your nesting of `TemplateDataProcessor`

```
public class TemplateDataProcessor<T> : DataProcessor<TemplateData<T>, TemplateDataProcessor<T>> { }
```

When I change the above declaration to this:

```
public class TemplateDataProcessor<T> : DataProcessor<TemplateData<T>, DataProcessor<TemplateData<T>, T>>{}
```

It worked fine in .NET 4.7.2 as well. See screenshot below:

![image](https://user-images.githubusercontent.com/7371750/58241245-3161a900-7d6a-11e9-85d3-bda6d53a61a1.png)

**Comment by Stamo-Gochev** [comment]
@sujayvsarma You are right, this fixes the issue. However, it is still interesting why passing the same class (`TemplateDataProcessor`) as a generic fails. Do you have any suggestions?

**Comment by sujayvsarma** [comment]
On the .NET Fiddle, I see that the error thrown is a `TimeLimitExceeded`. This is what got me to try changing the types of various things in your code snippet. Your `StackOverFlowException` would be related -- too many cyclical calls. 

I noticed that when I changed DataProcessor to a concrete class (by removing the keyword `abstract`), that fixed the problem too. I am not too sure here, but could it have something to do with the chain of abstract classes? I have never done anything like this before and so cannot really comment.

Also look at the IL generated for both versions (where you subclass `TemplateDataProcessor` from `DataProcessor` vs `DataProcessorBase`. The IL seems to be identical.

Tagging @madskristensen to weigh in?

**Comment by svick** [comment]
The repeating portion of the stack trace is:

Microsoft.CSharp.dll!Microsoft.CSharp.RuntimeBinder.SymbolTable.LoadSymbolsFromType(System.Type originalType)	
 	Microsoft.CSharp.dll!Microsoft.CSharp.RuntimeBinder.SymbolTable.GetConstructedType(System.Type type, Microsoft.CSharp.RuntimeBinder.Semantics.AggregateSymbol agg)	
 	Microsoft.CSharp.dll!Microsoft.CSharp.RuntimeBinder.SymbolTable.LoadSymbolsFromType(System.Type originalType)	
 	Microsoft.CSharp.dll!Microsoft.CSharp.RuntimeBinder.SymbolTable.GetConstructedType(System.Type type, Microsoft.CSharp.RuntimeBinder.Semantics.AggregateSymbol agg)	
 	Microsoft.CSharp.dll!Microsoft.CSharp.RuntimeBinder.SymbolTable.LoadSymbolsFromType(System.Type originalType)	
 	Microsoft.CSharp.dll!Microsoft.CSharp.RuntimeBinder.SymbolTable.GetConstructedType(System.Type type, Microsoft.CSharp.RuntimeBinder.Semantics.AggregateSymbol agg)	

Since the stack overflow also happens on .Net Core 2.0, but not on .Net Core 2.1, I think this got fixed on .Net Core by one of @JonHanna's changes in that time frame, possibly https://github.com/dotnet/corefx/pull/25191.

cc: @JonHanna, @vsadov

**Comment by Stamo-Gochev** [comment]
Any updates?

**Comment by Stamo-Gochev** [comment]
Any updates? It will be helpful if you can tell if this is a valid bug.
